src/backend/distributed/commands/table.c

```diff
@@ -1343,21 +1601,22 @@ InterShardDDLTaskList(Oid leftRelationId, Oid rightRelationId,
 	int taskId = 1;
 
 	/*
-	 * If the rightPartitionMethod is a reference table, we need to make sure
-	 * that the tasks are created in a way that the right shard stays the same
-	 * since we only have one placement per worker. This hack is first implemented
-	 * for foreign constraint support from distributed tables to reference tables.
+	 * If the right relation is a reference table and left relation is not a citus
+	 * local table, we need to make sure that the tasks are created in a way that
+	 * the right shard stays the same since we only have one placement per worker.
+	 * If left relation is a citus local table, then we don't need to populate
+	 * reference table shards as we will set foreign key only on reference table's
+	 * coordinator placement.
 	 */
-	if (rightPartitionMethod == DISTRIBUTE_BY_NONE)
+	if (IsReferenceTable(rightRelationId) && !IsCitusLocalTable(leftRelationId))
 	{
-		int rightShardCount = list_length(rightShardList);
+		int rightShardCount PG_USED_FOR_ASSERTS_ONLY = list_length(rightShardList);
 		int leftShardCount = list_length(leftShardList);
 
 		Assert(rightShardCount == 1);
 
 		ShardInterval *rightShardInterval = (ShardInterval *) linitial(rightShardList);
-		for (int shardCounter = rightShardCount; shardCounter < leftShardCount;
-			 shardCounter++)
+		for (int shardCounter = 1; shardCounter < leftShardCount; shardCounter++)
```

Comment:
It was really an implicit thing, it did take some time for me to understand too at first glance.
Here what happens is, as the left relation is a distributed table with `n` shards and right relation is a reference table with one shard (but placement at each node), we are populating reference table shard `rightShardList` n-1 times so we can iterate with `forboth` in below loop

Reply:
umm maybe create a function for this part:

```c
	/*
	 * If the right relation is a reference table and left relation is not a citus
	 * local table, we need to make sure that the tasks are created in a way that
	 * the right shard stays the same since we only have one placement per worker.
	 * If left relation is a citus local table, then we don't need to populate
	 * reference table shards as we will set foreign key only on reference table's
	 * coordinator placement.
	 */
	if (IsReferenceTable(rightRelationId) && !IsCitusLocalTable(leftRelationId))
	{
		int rightShardCount PG_USED_FOR_ASSERTS_ONLY = list_length(rightShardList);
		int leftShardCount = list_length(leftShardList);

		Assert(rightShardCount == 1);

		ShardInterval *rightShardInterval = (ShardInterval *) linitial(rightShardList);
		for (int shardCounter = 1; shardCounter < leftShardCount; shardCounter++)
		{
			rightShardList = lappend(rightShardList, rightShardInterval);
		}
	}
```

and add your finding as a comment there so that people don't need to spend time to understand that part.  A better thing would be to simplify this part if possible because what is done here doesn't look explicit.